### PR TITLE
vsr: make repair less eager

### PR DESCRIPTION
**Pull request by chaitanyabhandari**
Missing prepares/headers are expected during normal processing.
This is by virtue of experimental ops in adaptive replication routing
which cause replicas that usually receive ops in 2 hops to receive
*some* ops in 1 hop (and vice versa).

Currently, we do a lot of repair work (request_prepares, request_headers)
even during normal processing. This is because our repair is triggered
on *every prepare* we receive, so it is very eager to repair the 
aforementioned missing headers/prepares.

Now, if a missing prepare/header exists within a pipeline of ops
from the head, we wait for it to arrive via normal replication,
instead of eagerly repairing it. We use the distance between
our head op and the closest missing prepare/header as a proxy
for how much time has passed. Specifically, if we get >= pipeline
new prepares atop the missing prepare/header, it means that the
prepare may actually have been dropped/delayed on the network.

However, this distance may not increase in an idle cluster, so
we unconditionally trigger repair once every 50 journal_repair_timeout
attempts. Since journal_repair_timeout is between [75ms, 150ms], it
means we unconditionaly repair every [3.75s, 7.5s]. One caveat
of relying on attempts here is that we spuriously trigger
unconditional repair via other code paths (on_prepare, for example)
on every 50th journal_repair_timeout attempt.

---

### DPT results

<details>
  <summary><b>./zig/zig build -Drelease vopr -- --performance --requests-max=15000</b></summary>
<br>
Results over a 100 runs, massive reduction in `request_prepare`, `request_headers`, _and_ `prepares`, with only a minor, ~2% performance overhead:

Metric | Old | New | Avg Δ
-- | -- | -- | --
request_prepare | 73,334 | 10,177 | -86.1%
request_headers | 49,668 | 4,180 | -91.6%
headers | 44,725 | 3,759 | -91.6%
prepare | 153,819 | 102,337 | -33.5%
ticks | 94,348 | 96,431 | +2.2%

</details>

<details>
  <summary><b>./zig/zig build -Drelease vopr -- --performance --replica-missing=3 --replica-missing-until-request=7500 --requests-max=15000</b></summary>
<br>

Results over a 100 runs, there are 3 sets of buckets, in 23/100 runs `main` did some extra state sync (signaled by -50% Δ), and 19/100 runs where `cbb/repair_less_eager` did some extra state sync (signaled by +103.4% Δ), and finally 58/100 runs where there were no spurious state syncs by either, and block traffic was stable. The 3rd bucket, we see similar trends as above:

##### 1. Block traffic decreased (23/100)

Metric | Old | New | Avg Δ
-- | -- | -- | --
block | 29,762 | 14,722 | -50.5%
ticks | 141,180 | 130,223 | -7.8%

##### 2. Block traffic increased (19/100)

Metric | Old | New | Avg Δ
-- | -- | -- | --
block | 14,645 | 29,790 | +103.4%
ticks | 126,553 | 144,645 | +14.3%

##### 3. Block traffic stable (58/100)

Metric | Old | New | Avg Δ
-- | -- | -- | --
request_prepare | 61,030 | 19,508 | -68.0%
request_headers | 35,708 | 4,681 | -86.9%
headers | 27,709 | 3,608 | -87.0%
prepare | 133,767 | 104,314 | -22.0%
block | 15,069 | 15,036 | -0.2%
ticks | 126,163 | 129,796 | +2.9%

</details>

### CODE REVIEW

File: src/vsr/replica_test.zig [hunk 1]
```diff
@@ -2269,7 +2269,7 @@ const TestContext = struct {
     }
 
     pub fn run(t: *TestContext) void {
-        const tick_max = 4_100;
+        const tick_max = 8_200;
```
Comment (chaitanyabhandari) [inline src/vsr/replica_test.zig:2272]: Required as in an idle cluster, we only _unconditionally_ trigger repair once every 50 `journal_repair_timeout` invocations.

File: src/vsr/replica.zig [hunk 2]
```diff
@@ -7570,48 +7570,73 @@ pub fn ReplicaType(
...
+            // cause replicas that usually receive ops in 2 hops to
+            // receive *some* ops in 1 hop (and vice versa). So, if a
+            // missing prepare/header exists within a pipeline of ops
+            // from the head, wait for it to arrive via normal
+            // replication, instead of eagerly repairing it.
+            for (self.op -| constants.pipeline_prepare_queue_max..self.op + 1) |op| {
+                // Every 50 timeouts, unconditionally repair. This allows
+                // backups to repair journal faults in an idle cluster where
+                // the the head op does not progress.
+                if (self.journal_repair_timeout.attempts % 50 == 0) continue;
```
Comment (chaitanyabhandari) [inline src/vsr/replica.zig]: Would appreciate thoughts on this particular statement we had to add to deal with idle clusters that have a missing op within a pipeline of the head op.

Another alternative I considered here is that the cluster keeps sending `pulse` ops through the log even when the cluster is idle. However, that seems like a _major_ departure from our current design, where we only pulse when needed by the state machine (when there's a transfer with an expiry), which is quite clean. Also, I think adding these unconditional pulses would also break our replica tests, as none of the asserts about commit & op would continue to hold (as the cluster internally keeps injecting pulses), which is quite bad.

Comment (sentientwaffle) [reply src/vsr/replica.zig]: 100% agree -- if for no other reason than "idle cluster doesn't commit" is also a much simpler mental model!

File: src/vsr/replica.zig [hunk 3]
```diff
@@ -7570,48 +7570,73 @@ pub fn ReplicaType(
...
-            );
+            // Missing prepares/headers are expected during normal
+            // processing. This is by virtue of experimental ops in
+            // adaptive replication routing (see routing.zig), which
+            // cause replicas that usually receive ops in 2 hops to
+            // receive *some* ops in 1 hop (and vice versa). So, if a
+            // missing prepare/header exists within a pipeline of ops
+            // from the head, wait for it to arrive via normal
+            // replication, instead of eagerly repairing it.
+            for (self.op -| constants.pipeline_prepare_queue_max..self.op + 1) |op| {
```
Comment (sentientwaffle) [inline src/vsr/replica.zig]: Could we refactor this to simplify the control flow? Right now we are using the `for` loop as a complex condition.

I think something like this would be easier to follow:

```zig
const repair_header_breaks: repair: {
    if (self.journal_repair_timeout.attempts % 50 == 0) break :repair true;
    if (self.status == .view_change) break :repair true;
    for (...) |op| {
        if (slot...) {
             break :repair false;
        }
    }
    break :repair true;
};

if (repair_header_breaks) {
    const header_break = ...
}
```

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig:1]: Just to sanity-check, can you also compare before/after of performance-vopr with replica-missing-until to check behavior during state sync?

Comment (chaitanyabhandari) [reply src/vsr/replica.zig:1]: Ah, I was also thinking about this _just_ this morning! Coming up shortly @sentientwaffle. :)

File: src/vsr/replica.zig (hunk 3 above)
Comment (chaitanyabhandari) [reply src/vsr/replica.zig]: Great point DJ! I actually ended simplifying it even further, getting rid of the for loop entirely, and instead repairing up to `self.op -| constants.pipeline_prepare_queue_max` in the absence of the first two conditions:
```C
            const repair_op_max: u64 = repair: {
                // Every 50 timeouts, unconditionally repair. This
                // allows backups to repair journal faults in an idle
                // cluster where the the head op does not progress.
                if (self.journal_repair_timeout.attempts % 50 == 0) break :repair self.op;

                // View changing replicas must unconditionally repair,
                // as transitioning to normal status requires them to
                // repair their journal.
                if (self.status == .view_change) break :repair self.op;

                // Missing prepares/headers are expected during normal
                // processing. This is by virtue of experimental ops in
                // adaptive replication routing (see routing.zig), which
                // cause replicas that usually receive ops in 2 hops to
                // receive *some* ops in 1 hop (and vice versa). So, if a
                // missing prepare/header exists within a pipeline of ops
                // from the head, wait for it to arrive via normal
                // replication, instead of eagerly repairing it.
                break :repair self.op -| constants.pipeline_prepare_queue_max;
            };
```

File: src/vsr/replica.zig [hunk 4]
```diff
@@ -8173,49 +8210,13 @@ pub fn ReplicaType(
...
 
-        fn repair_prepares(self: *Replica) void {
+        fn repair_clean_out_of_bound_prepares(self: *Replica) void {
             assert(self.status == .normal or self.status == .view_change);
             assert(self.repairs_allowed());
             maybe(self.journal.dirty.count > 0);
             assert(self.op >= self.commit_min);
             assert(self.op - self.commit_min <= constants.journal_slot_count);
 
-            if (self.op < constants.journal_slot_count) {
```
Comment (chaitanyabhandari) [inline src/vsr/replica.zig:8183]: IIUC, this first if is subsumed by the second one. Specifically, if we are in the first wrap, in the second if block, we truncate dirty slots _not_ between [0, self.op], which is what we _also_ do in the first if block. Hope I'm not missing anything here! :)

File: src/vsr/replica.zig [hunk 5]
```diff
@@ -7593,25 +7615,40 @@ pub fn ReplicaType(
                         self.op,
                     },
                 );
+                const op_min = @min(range.op_min, self.repair_header_op_next);
+                const op_max = @min(repair_op_max, range.op_max);
+                if (op_min <= op_max) {
```
Comment (sentientwaffle) [inline src/vsr/replica.zig]: ~Hmm, this is tricky. In the case that `repair_op_max == self.op -| constants.pipeline_prepare_queue_max`, it may be that `op_max == repair_op_max`. Then we request the headers from `op_min ... repair_op_max`. But we might not have `repair_op_max` in our journal (it is part of the pipeline, so we might not be repairing it yet). In which case, we will ignore the incoming headers that we just requested, since they don't hash-chain.~

Ahh wait, `repair_header` has a fast-path when the view matches, so we are actually fine in this case!

**Comment by chaitanyabhandari** [comment]
Results from `./zig/zig build -Drelease vopr -- --performance --replica-missing=3 --replica-missing-until-request=7500 --requests-max=15000` up in the PR description!

File: src/vsr/replica.zig [hunk 6]
```diff
@@ -7581,37 +7598,51 @@ pub fn ReplicaType(
...
-                        range.op_min,
-                        range.op_max,
-                        self.commit_min,
-                        self.commit_max,
-                        self.op,
-                    },
-                );
+                if (range.op_min <= repair_op_max) {
+                    const op_min = range.op_min;
+                    const op_max = @min(range.op_max, repair_op_max);
```
Comment (sentientwaffle) [inline src/vsr/replica.zig:7603]: ~Hmm, this is tricky. In the case that `repair_op_max == self.op -| constants.pipeline_prepare_queue_max`, it may be that `op_max == repair_op_max`. Then we request the headers from `op_min ... repair_op_max`. But we might not have `repair_op_max` in our journal (it is part of the pipeline, so we might not be repairing it yet). In which case, we will ignore the incoming headers that we just requested, since they don't hash-chain.~

Ahh wait, `repair_header` has a fast-path when the view matches, so we are actually fine in this case!

Comment (sentientwaffle) [inline src/vsr/replica.zig:7603]: Let's add `assert(op_min <= op_max);`

File: src/vsr/replica.zig
Comment (sentientwaffle) [inline src/vsr/replica.zig]: Could write this as:
```zig
for (0..constants.journal_slot_count) |slot_index| {
```

Comment (sentientwaffle) [inline src/vsr/replica.zig]: Comment out of date, there's no previous loop anymore.

File: src/vsr/replica.zig (hunk 4 above)
Comment (sentientwaffle) [reply src/vsr/replica.zig:8183]: Hmm, suppose `self.op == 0`, and some `op > 0` is faulty. Then `slots_repaired.head == slots_repaired.tail`, but we wouldn't truncate anything, right? Whereas before the first loop would have truncated everything beyond op `0`. That is the only case I can think of where behavior might differ, though.

File: src/vsr/replica.zig (hunk 6 above)
Comment (chaitanyabhandari) [reply src/vsr/replica.zig:7603]: Yes, I think the commit & repair fast path for when the head op's view is the same as the one that we're repairing or committing would save us here!

File: src/vsr/replica.zig (hunk 4 above)
Comment (chaitanyabhandari) [reply src/vsr/replica.zig:8183]: Mhmmmmmm that's a great one DJ, but then in that case we will never try to read such a slot, and we just overwrite it with a non-zero op, no?

Comment (sentientwaffle) [reply src/vsr/replica.zig:8183]: Yep, I think it is fine!

File: src/vopr.zig [hunk 7]
```diff
@@ -282,10 +288,22 @@ pub fn main() !void {
...
+            simulator.options.cluster.replica_count,
+            simulator.options.cluster.standby_count,
+        ));
+
+        tick = 0;
+        while (tick < cli_args.ticks_max_convergence) : (tick += 1) {
+            simulator.tick();
+            tick_total += 1;
+            if (simulator.pending() == null) break;
+        }
```
Comment (chaitanyabhandari) [inline src/vopr.zig:311]: Earlier, we were choosing a `random_core` for performance mode but perhaps better to just wait for the _entire_ cluster to converge?

Comment (sentientwaffle) [reply src/vopr.zig:311]: Good point -- though we should exclude the `replica_missing`.

File: src/vopr.zig [hunk 8]
```diff
@@ -832,7 +850,7 @@ pub const Simulator = struct {
         if (simulator.options.replica_missing_until_request) |request| {
             if (simulator.requests_replied >= request) {
                 simulator.options.replica_missing_until_request = null;
-                simulator.replica_crash_stability[simulator.options.replica_missing.?] = 1;
```
Comment (chaitanyabhandari) [inline src/vopr.zig:835]: This seems like a bug to me that would cause us to _always_ restart this missing replica during the _end_ of the VOPR run, during `transition_to_liveness_mode`.  Even though we set `replica_crash_stability=0`, which would cause us to call `tick_crash_down` for this replica:
```C
    fn tick_crash(simulator: *Simulator) void {
        for (simulator.cluster.replicas) |*replica| {
            simulator.replica_crash_stability[replica.replica] -|= 1;
            if (simulator.replica_crash_stability[replica.replica] > 0) continue;

            switch (simulator.cluster.replica_health[replica.replica]) {
                .up => |up| {
                    if (!up.paused) simulator.tick_crash_up(replica);
                },
                .down => simulator.tick_crash_down(replica),
                .reformatting => {},
            }
        }
    }
```

`replica_restart_probability=0` for performance mode, so we don't actually get to the `restart_replica` call that follows:
```C
        const restart_random =
            simulator.prng.chance(simulator.options.replica_restart_probability);
        if (!restart_upgrade and !restart_random) return;

```

File: src/vopr.zig (hunk 7 above)
Comment (chaitanyabhandari) [reply src/vopr.zig:311]: Oh now, we also assert that `replica_missing` will come up _before_ `requests_max` are replied to, so by the time we reach this it should _already_ be up. We can assert that actually! :D 

```C
    if (options.replica_missing_until_request != null and
        options.requests_max < options.replica_missing_until_request.?)
    {
        return vsr.fatal(.cli, "--requests-max < --replica-missing-until-request", .{});
    }
```

Comment (chaitanyabhandari) [reply src/vopr.zig:311]: Ah, but I think it is perhaps valuable to _also_ let the replica be down for entirety of the run, which is  if `--replica-missing-until-request` is omitted, so _indeed_ we should exclude replica missing from here, ` cli_args.replica_missing_until_request == null and  cli_args.replica_missing != null`.

**Review by sentientwaffle — APPROVED** [review]
LGTM, though let's let it bake in CFO for a bit before merging!

**Comment by chaitanyabhandari** [comment]
Green at `937 409` vopr-lite & `674 647` vopr! Merging this.
Screenshot 2026-04-16 at 7 08 15 AM
